File: internal/lua_engine/sqli_boolean.go
package lua_engine

import (
	"bytes"
	"context"
	"fmt"
	"net/http"

	"github.com/londonmax12/hyperz/internal/httpclient"
)

// SQLiBoolean probes for SQL injection through differential response
// analysis: for each sink we send a baseline request, a truthy probe
// (`<value>' AND '1'='1`), and a falsy probe (`<value>' AND '1'='2`),
// then ask BooleanCompare whether truthy~baseline and falsy!=baseline.
// A vulnerable parameter behaves like a SQL boolean - truthy reproduces
// the baseline row set, falsy collapses to an empty one - while a
// well-parameterized input ignores the trailing payload entirely (both
// probes look like baseline).
//
// Before handing bodies to the oracle we strip the literal pair suffix
// from each variant body so a page that simply echoes its input value
// doesn't artificially diverge truthy from falsy. The baseline is
// untouched (it never carried a pair suffix to begin with), so the
// three-way comparison stays fair on echo-heavy pages.
//
// Request count per sink: 1 baseline + 2*N pair probes, stopping at the
// first pair the oracle classifies as vulnerable. With 4 curated pairs
// the worst case is 9 requests per sink.
//
// Active (LevelDefault) check.
type SQLiBoolean struct{}

// sqliBooleanBodyCap bounds the response body the check reads. Larger
// than the SQLi-error cap because the oracle's similarity scoring needs
// enough of the templated page to be comparable: a 4 KiB sample of a
// 60 KiB dashboard wouldn't capture the row-set divergence we're after.
const sqliBooleanBodyCap = 64 << 10

// probe runs the baseline + truthy/falsy sweep for one sink. Returns a
// finding only when BooleanCompare verdicts BoolVulnerable; BoolNoSignal
// and BoolIndeterminate are both treated as "no actionable evidence" so
// the check stays high-precision at the cost of missing parameters whose
// truthy/falsy split doesn't cleanly match the baseline pattern.
func (c SQLiBoolean) probe(ctx context.Context, client *httpclient.Client, target string, sink Sink) (*Finding, error) {
	baseValue := sink.Value
	_, baseResp, baseBody, _, err := c.send(ctx, client, sink, baseValue)
	if err != nil {
		return nil, err
	}
	baseSnap := Snapshot{Status: statusOf(baseResp), Body: baseBody}

	for _, pair := range SQLiBooleanPairs() {
		if ctx.Err() != nil {
			break
		}
		truthyWire := baseValue + pair.True
		falsyWire := baseValue + pair.False

		// Pair-level send errors do not disqualify the sink: one flaky
		// request only invalidates this pair's verdict, not the next
		// pair's. Report and continue so a transient blip on the
		// string-quoted pair doesn't hide a vulnerability the numeric
		// pair would otherwise expose.
		_, tResp, tBody, _, err := c.send(ctx, client, sink, truthyWire)
		if err != nil {
			Report(ctx, fmt.Errorf("sqli-boolean truthy %s %s=%s pair=%s: %w",
				sink.Loc, sink.Name, sink.URL, pair.Name, err))
			continue
		}
		fReq, fResp, fBody, fTruncated, err := c.send(ctx, client, sink, falsyWire)
		if err != nil {
			Report(ctx, fmt.Errorf("sqli-boolean falsy %s %s=%s pair=%s: %w",
				sink.Loc, sink.Name, sink.URL, pair.Name, err))
			continue
		}

		// Strip the literal pair suffix from each variant body before
		// comparison: if the app echoes its input, leaving the suffix in
		// place would make every truthy/falsy body diverge from baseline
		// regardless of SQL behavior, producing a false positive on any
		// reflecting page.
		tStripped := bytes.ReplaceAll(tBody, []byte(pair.True), nil)
		fStripped := bytes.ReplaceAll(fBody, []byte(pair.False), nil)

		result := BooleanCompare(
			baseSnap,
			Snapshot{Status: statusOf(tResp), Body: tStripped},
			Snapshot{Status: statusOf(fResp), Body: fStripped},
		)
		if result.Decision != BoolVulnerable {
			continue
		}

		probeURL := ""
		method := ""
		if fReq != nil {
			method = fReq.Method
			if fReq.URL != nil {
				probeURL = fReq.URL.String()
			}
		}
		status := statusOf(fResp)
		return &Finding{
			Check:    "sqli-boolean",
			Target:   target,
			URL:      probeURL,
			Severity: SeverityHigh,
			Title:    fmt.Sprintf("SQL injection (boolean-based) in %s parameter %q", sink.Loc, sink.Name),
			Detail: fmt.Sprintf(
				"Parameter %q (%s) responds to SQL boolean inference: pair sqli-boolean/%s produced "+
					"truthy~baseline (sim=%.3f, status=%d) and falsy!=baseline (sim=%.3f, status=%d). "+
					"%s. An attacker can extract database contents by chaining boolean conditions.",
				sink.Name, sink.Loc, pair.Name,
				result.TruthySim, statusOf(tResp), result.FalsySim, status, result.Detail),
			CWE:   "CWE-89",
			OWASP: "A03:2021 Injection",
			Remediation: "Use parameterized queries / prepared statements so user input is bound as a value, never " +
				"concatenated into SQL text. Boolean-based SQLi remains exploitable even when verbose errors are disabled, " +
				"so suppressing error output alone is not a fix.",
			Evidence: &Evidence{
				Method:     method,
				RequestURL: probeURL,
				Status:     status,
				Snippet:    snippet(fBody, []byte(pair.False), false),
				Exchange:   RecordExchange(fReq, nil, false, fResp, fBody, fTruncated),
			},
			DedupeKey: MakeKey("sqli-boolean", ScopeParam, target, "loc:"+string(sink.Loc), "param:"+sink.Name),
		}, nil
	}
	return nil, nil
}

// send mutates sink with wireValue, dispatches the request, and reads up
// to sqliBooleanBodyCap of the body. Mirrors the sibling checks' send
// shape so a future shared HTTP shell drops in without per-check change.
func (c SQLiBoolean) send(ctx context.Context, client *httpclient.Client, sink Sink, wireValue string) (*http.Request, *http.Response, []byte, bool, error) {
	req, err := sink.MutateRequest(ctx, wireValue)
	if err != nil {
		return nil, nil, nil, false, err
	}
	resp, err := client.Do(ctx, req)
	if err != nil {
		return req, nil, nil, false, err
	}
	defer resp.Body.Close()
	body, truncated, err := httpclient.ReadBodyCapped(resp, sqliBooleanBodyCap)
	if err != nil {
		return req, resp, nil, false, err
	}
	return req, resp, body, truncated, nil
}

File: internal/checks/sqli_boolean.lua
-- sqli-boolean: per sink, send a baseline plus a truthy/falsy pair
-- sweep. ctx.oracle.boolean_compare decides whether the differential
-- evidence (truthy ~ baseline AND falsy != baseline) is vulnerability-
-- shaped. Per-pair bodies are stripped of the literal pair suffix
-- before the oracle runs so an echo-heavy page doesn't artificially
-- diverge truthy from falsy.

local check = {
  name        = "sqli-boolean",
  level       = "default",
  scope       = "param",
  cwe         = "CWE-89",
  owasp       = "A03:2021 Injection",
  remediation = "Use parameterized queries / prepared statements so user input is bound as a value, never "
                .. "concatenated into SQL text. Boolean-based SQLi remains exploitable even when verbose errors are disabled, "
                .. "so suppressing error output alone is not a fix.",
  consumes    = {"page", "param"},
}

local BODY_CAP = 64 * 1024

local function send(ctx, sink, wire_value)
  local req, mut_err = sink:mutate_request(wire_value)
  if mut_err then return nil, nil, nil, false, mut_err end
  local resp, do_err = ctx.client["do"](ctx.client, req)
  if do_err then return req, nil, nil, false, do_err end
  local body, truncated, rerr = resp:read_body_capped(BODY_CAP)
  if rerr then return req, resp, nil, false, rerr end
  return req, resp, body, truncated, nil
end

-- strip_all removes every occurrence of needle from body. gopher-lua's
-- gsub treats the pattern as a Lua pattern by default; we want a
-- literal substring scrub so callers can pass arbitrary SQL fragments
-- without escaping each pattern metacharacter. plain-find + concat keeps
-- the scrub literal.
local function strip_all(body, needle)
  if body == "" or needle == "" then return body end
  local out, i = {}, 1
  while true do
    local s, e = string.find(body, needle, i, true)
    if not s then
      out[#out + 1] = string.sub(body, i)
      break
    end
    out[#out + 1] = string.sub(body, i, s - 1)
    i = e + 1
  end
  return table.concat(out)
end

local function probe(ctx, sink)
  local base_value = sink.value
  local _, base_resp, base_body, _, base_err = send(ctx, sink, base_value)
  if base_err then return nil, base_err end
  local base_status = base_resp and base_resp:status() or 0
  local baseline_snap = { status = base_status, body = base_body }

  for _, pair in ipairs(ctx.payloads.sqli_boolean_pairs()) do
    local truthy_wire = base_value .. pair.truthy
    local falsy_wire = base_value .. pair.falsy

    -- Pair-level send errors do not disqualify the sink: one flaky
    -- request only invalidates this pair's verdict, not the next
    -- pair's.
    local _, t_resp, t_body, _, t_err = send(ctx, sink, truthy_wire)
    if t_err then
      ctx:report(string.format("sqli-boolean truthy %s %s=%s pair=%s: %s",
        sink.loc, sink.name, sink.url, pair.name, t_err))
    else
      local f_req, f_resp, f_body, f_truncated, f_err = send(ctx, sink, falsy_wire)
      if f_err then
        ctx:report(string.format("sqli-boolean falsy %s %s=%s pair=%s: %s",
          sink.loc, sink.name, sink.url, pair.name, f_err))
      else
        local t_stripped = strip_all(t_body, pair.truthy)
        local f_stripped = strip_all(f_body, pair.falsy)
        local t_status = t_resp and t_resp:status() or 0
        local f_status = f_resp and f_resp:status() or 0

        local result = ctx.oracle.boolean_compare(
          baseline_snap,
          { status = t_status, body = t_stripped },
          { status = f_status, body = f_stripped })

        if result.decision == "vulnerable" then
          local probe_url = f_req:url()
          return {
            severity = ctx.severity.high,
            url      = probe_url,
            title    = string.format('SQL injection (boolean-based) in %s parameter "%s"', sink.loc, sink.name),
            detail   = string.format(
              'Parameter %q (%s) responds to SQL boolean inference: pair sqli-boolean/%s produced '
                .. 'truthy~baseline (sim=%.3f, status=%d) and falsy!=baseline (sim=%.3f, status=%d). '
                .. '%s. An attacker can extract database contents by chaining boolean conditions.',
              sink.name, sink.loc, pair.name,
              result.truthy_sim, t_status, result.falsy_sim, f_status, result.detail),
            evidence = ctx.evidence.from_exchange {
              request   = f_req,
              response  = f_resp,
              body      = f_body,
              truncated = f_truncated,
            },
            dedupe_parts = { "loc:" .. sink.loc, "param:" .. sink.name },
          }
        end
      end
    end
  end
  return nil
end

function check.run(ctx)
  local u, perr = ctx.url.parse(ctx.page.url)
  if perr or not u or u.scheme == "" or u.host == "" then return nil end
  if not ctx.scope:allows(ctx.page.url) then return nil end

  local sinks = ctx.sinks.for_page{}
  if #sinks == 0 then return nil end

  local findings = {}
  local seen = {}
  local first_err
  local probed_any = false
  for _, sink in ipairs(sinks) do
    if ctx.scope:allows(sink.url) then
      local f, err = probe(ctx, sink)
      if err then
        ctx:report(string.format("probe %s %s=%s: %s", sink.loc, sink.name, sink.url, err))
        if not first_err then first_err = err end
      else
        probed_any = true
        if f then
          local key = "loc:" .. sink.loc .. "|param:" .. sink.name
          if not seen[key] then
            seen[key] = true
            findings[#findings + 1] = f
          end
        end
      end
    end
  end

  if not probed_any and first_err then return nil, first_err end
  return findings
end

return check
